crates/ra_hir/src/ty.rs

```diff
@@ -973,14 +1143,24 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
                 ret_ty
             }
             Expr::Match { expr, arms } => {
-                let _ty = self.infer_expr(*expr, &Expectation::none());
-                for arm in arms {
-                    // TODO type the bindings in pats
+                let mut expected = expected.clone();
```

Comment:
instead of changing the expected type with each match arm, I think it'd be better to use a new type variable here if `expected.ty` is unknown; then you don't need to reassign `expected` at all in the loop below